### ISSUE TITLE
Allow concurrent creation of SecretBinding and Shoots using the binding

### ISSUE THREAD

**Issue by loewenstein** [gardener/gardener#1896]
**What would you like to be added**:
I would like to be able to create Shoots referencing SecretBindings that do not exist yet. Similar to a Pod that is still missing volume mounts, the Shoot controller should watch for the SecretBinding to be available and start Shoot provisioning eventually.

**Why is this needed**:
We want to provide the infrastructure secrets and trigger the Shoot creation without orchestrating this as sequential actions. E.g. have a single Helm Chart that will create the Shoot and SecretBinding.

**Comment by vpnachev** [comment]
Can we even extend this further?
I mean to make the SecretBinding optional and if it is not provided:
* Gardener will provision only the Control Plane and will not take care of the infrastructure behind the shoot cluster, e.g. VPC, subnet, VMs. 
* If a secret binding is provided later, Gardener will take care to provision the omitted resources.
* I have to think whether it should be possible to remove the secret binding from an existing cluster with already provisioned infrastructure.

Generally, such change seems simple, but it will require a lot of fields in the shoot spec to be made optional, too - like the `shoot.spec.provider.[infrastructureConfig|workers]`.

**Comment by loewenstein** [comment]
@vpnachev It would be great if the implementation of the basic use case will not be blocked on a discussion of the extended use cases.

**Comment by vpnachev** [comment]
As long as all of this is exposed in public API that is already in beta version, I would like to have a proper design so that there will be no braking changes in the future.

But generally, we have to find out whether it makes sense to enable the option CPaaS(Control Plane as a service) next to KaaS(Kubernetes as a Service) via Gardener.

@vlerenc @marwinski @mvladev @rfranzke @vasu1124 - I would like to hear your opinion about the option for CP only? Now, I start thinking that if Gardener will support such use case, it should be via dedicated resource and not to re-use the Shoot.

**Comment by vpnachev** [comment]
After one more night thinking about this, I think it will not be nice to use dedicated resource for CPaaS, because it will be hard to offer automatic switching between CPaaS and K8SaaS for existing clusters.

The shoot resource can be re-used in the following way:
* if `shoot.spec.provider.infrastructureConfig` and `shoot.spec.provider.workers` are omitted, this means that shoot cluster is limited only to its control plane. In this case `shoot.spec.secretBindingName` is not required.

* if `shoot.spec.provider.infrastructureConfig` and `shoot.spec.provider.workers` are provided, this means that shoot cluster is fully featured K8S cluster. In this case `shoot.spec.secretBindingName` is required. If it is not provided, the reconciliation can hangs till it is created, just like it is waiting Gardener Scheduler to schedule the shoot on appropriate seed cluster. Note, as long as the Infrastructure and Workers exists, the secret binding cannot be removed. In this way, the migration from K8S to CP only is possible.

With this said, I think there is no blocker to make the secret binding optional and let reconciliation wait till it is provided.

/cc @ialidzhikov for feedback.

**Comment by rfranzke** [comment]
@vpnachev I think this is not an easy discussion and clearly out of the scope of this issue. I suggest to create a different issue/GEP for this discussion.
Though, generally, please keep in mind that the control plane also contains provider-specific components like cloud-controller-manager or CSI controllers which are requiring a) the cloud provider credentials and b) proper configuration. I don't think it's reasonable to offer CPaaS without these essential components. If you want to offer CPaaS you have to find a way to let the user configure these components for his use-case (we don't control the infrastructure anymore, hence, user needs full control). Other than that, generally, I think CPaaS could be interesting for some use-cases (I remember being asked about this several times at KubeCons). Also, it's more or less the model that EKS is implementing.

**Comment by vpnachev** [comment]
> @vpnachev I think this is not an easy discussion and clearly out of the scope of this issue. I suggest to create a different issue/GEP for this discussion.

Sure, I just wanted to understand whether making optional `secretBindingName` can affect future features based on the Shoot definitions. In my last assumption, it is fine to do it.

I will follow up with another issue collecting the info from this one, which can be used for a GEP about CPaaS only.
